### ISSUE TITLE
Remove -std=c++11 flag from CMakeLists.txt

### ISSUE THREAD

**Issue by tdalman** [open-source-parsers/jsoncpp#548]
Compiling jsoncpp with a fairly old cross-gcc (4.6) fails for me, because it does not support -std=c++11. Instead, I would have to set -std=c++0x.

With CMake, rather than specifying  the -std= flag, [CXX_EXTENSIONS](https://cmake.org/cmake/help/v3.1/prop_tgt/CXX_EXTENSIONS.html) automatically enables C++11 if available (which is on by default).

My proposal is to remove -std=c++11 entirely from CMakeLists.txt. I can provide a patch if needed.
Also, checks for required [C++11 compiler features](https://cmake.org/cmake/help/v3.1/manual/cmake-compile-features.7.html#manual:cmake-compile-features(7)) could be added to further improve portability.

**Comment by cdunn2001** [comment]
Personally, I have no strong opinion. We are definitely using **CMake**. If `CXX_EXTENSIONS` is a several-year old feature of cmake, then I don't see a reason not to use it. However, it is nice to require C++11 in advance, so nobody can complain about future changes.

Any other opinions?

**Comment by tdalman** [comment]
How about we add checks for the particular C++11 features we actually use now? Otherwise, full C++11 is only supported with GCC >= 4.8 and MS Visual Studio >= 2015.

**Comment by cdunn2001** [comment]
Does `c++0x` definitely work for you? And passes all tests? If so, please submit a PR for `CXX_EXTENSIONS`.

**Comment by strahlc** [comment]
If you have cmake 3.1 or higher, you can use 

``` cmake
cmake_minimum_required(VERSION 3.1)
set(CMAKE_CXX_STANDARD 11)
set(CMAKE_CXX_STANDARD_REQUIRED ON)
```
instead. 
This will enable at least C++11 (or higher if available).

**Comment by cdunn2001** [comment]
We still need to remove the current `--std` flag, right? PR?

**Comment by strahlc** [comment]
Yes, you'll have to remove the --std=... flag. cmake will take care of it.

**Comment by SoapGentoo** [comment]
In me refactoring in https://github.com/open-source-parsers/jsoncpp/pull/562 I've already done this. I believe requiring CMake 3.1 is not too much of a burden nowadays.

**Comment by cdunn2001** [comment]
We're a bit limited by TravisCI. See the build failure:

> CMake Error: INSTALL(EXPORT "jsoncpp") given absolute DESTINATION "/jsoncpp" but the export references an installation of target "jsoncpp_lib" which has relative DESTINATION "lib".

**Comment by tdalman** [comment]
I wasn"t aware of CMAKE_CXX_STANDARD. The approach works for me and I think we should go this path.

The install(EXPORT ...) problem seems to be a different issue that needs fixing before we can close this one. Christopher, have you created a ticket already?

**Comment by SoapGentoo** [comment]
@cdunn2001 I fixed some missing parts, but due to the CMake 3.1 requirement, Travis fails now. What should I do?

**Comment by Gachapen** [comment]
I've added some features that require a newer version of CMake, and fixed the issue by adding a version check that only uses the feature if it is available and falls back otherwise. See for example this https://github.com/open-source-parsers/jsoncpp/blob/master/src/lib_json/CMakeLists.txt#L20 or this https://github.com/open-source-parsers/jsoncpp/blob/master/src/lib_json/CMakeLists.txt#L87

The alternative is to somehow upgrade the Travis CMake (I'm all for this).

**Comment by SoapGentoo** [comment]
I think this issue can be closed now, seeing how my latest change requires ISO (and not GNU) C++ and bumps the CMake requirement to 3.1.

**Comment by cdunn2001** [comment]
Agreed. But if anyone runs into a problem, feel free to re-open.
